### ISSUE TITLE
FSDP crash if no parameters are used in fwd pass

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#83107]
### 🐛 Describe the bug

The following test raises an issue where an assert is triggered if FSDP has params that require grad, but they are not used in an iteration:

```
@skip_if_lt_x_gpu(2)
    def test_fsdp_namedtuple(self):

        class MyModule(nn.Module):
            def __init__(self):
                super().__init__()
                self.lin = nn.Linear(1, 1)

            def forward(self, x):
                return x

        m = MyModule().cuda()
        m = FSDP(m)
        t = torch.ones(1, device="cuda", requires_grad=True)

        MyOutputType = namedtuple(
            "MyOutputType",
            ["a", "b", "c", "d"],
            defaults=(t, t, t, t)
        )

        inp = MyOutputType()
        out = m(inp)
        print(out)
        res = torch.cat([e for e in out]).sum()
        res.backward()
```

This triggers the assert here: https://github.com/pytorch/pytorch/blob/cd5efc6f082c81fd40712127638931b9e2e5ee69/torch/distributed/fsdp/fully_sharded_data_parallel.py#L3084, presumably because the FSDP managed param requires grad, but does not get its gradient computed, so it never entered post backward.

In practice, this came up when wrapping FLAVA encoders separately, but encoders take turns being used across iterations.

### Versions

main

cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501
